### PR TITLE
[SRVCOM-1532] Add the Troubleshooting Serverless section

### PR DESCRIPTION
**Pull request by rh-max**
Version(s): 4.6+

Issue: https://issues.redhat.com/browse/SRVCOM-1532

Link to docs preview: https://deploy-preview-45262--osdocs.netlify.app/openshift-enterprise/latest/serverless/reference/troubleshooting.html

### CODE REVIEW

File: modules/serverless-troubleshooting-knative-kafka.adoc [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+Use the web console, the Knative CLI, or a YAML to create a channel.
+
+*Kafka Event source is not visible after installing {ServerlessOperatorName}.*
+
+Make sure to do the following:
+
+* Install the `KnativeKafka` custom resource (CR).
+* Toggle the `enabled` switch for the options you want to use to `true`.
+* Specify the bootstrap servers to use the Kafka channel or broker.
```
Comment (mgencur) [inline modules/serverless-troubleshooting-knative-kafka.adoc:19]: This last bullet is not related to Kafka source. It's only related to Kafka channel and broker. It should be removed. IMO

File: modules/serverless-troubleshooting-monitoring-and-logging.adoc [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+//
+// * serverless/reference/troubleshooting.adoc
+
+:_content-type: REFERENCE
+[id="serverless-troubleshooting-monitoring-and-logging_{context}"]
+= Monitoring and logging
+
+*How can I access logs of my {ServerlessProductShortName} containers?*
+
+Container logs are available using the integration with {product-title} logging.
```
Comment (mgencur) [inline modules/serverless-troubleshooting-monitoring-and-logging.adoc:11]: I suppose after reading this sentence the user is supposed to look at this doc? https://github.com/openshift/openshift-docs/pull/45262/files#diff-17323afbc27eee94cd231ac0ad14a1d1b8c2278b059388b45db3db0d6d438fddR47 
It's not really clear where the user should go from here.

**Review by mgencur — CHANGES_REQUESTED** [review]
Hi @rh-max , this troubleshooting guide depends a lot on the content in "Additional resources". But it's not immediately clear where the user should look after reading individual sections.
Perhaps it would make sense to include more details in the individual section of this troubleshooting guide or at least clearly point to docs that the user should read to get more details.

File: modules/serverless-troubleshooting-general.adoc [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+:_content-type: REFERENCE
+[id="serverless-troubleshooting-general_{context}"]
+= General troubleshooting
+
+*How many {product-title} resources does {ServerlessProductShortName} Serving consume?*
+
+The request of the `knative-serving` component is 0.61 CPU and 400 megabytes of memory.
+
+The limit of the `knative-serving` component is 4.1 CPU and 4.71 gigabytes of memory.
```
[the remark this answers is not in the document]
Comment (rh-max) [reply modules/serverless-troubleshooting-general.adoc:13]: @mgencur Can you please provide the following:
* In `(300m + 100m + 100m + 30m + 100m) * 2`, what do the individual constituents stand for?
  * Or, if it is not important, can we just write `630m for every replica, resulting in 1260m for the default two replicas`?
* What are the numbers of Kourier-related pods?
  * Or should we just state `plus the requests of the Kourier-related pods`?
* What are the memory _limits_ for Knative Serving components and for Kourier-related pods?

**Comment by rh-max** [comment]
Hi @mgencur ! Thanks you, very valid feedback. I can not implement it optimally, because we can only link from the end of subsections.
However, I unified the section to always reference, as in-place plaintext, which section the user needs to look at.
I've removed some of the details I was previously referring to, because they are in the linked sections anyway, and we want to avoid duplication.
My remaining ask for you is regarding Knative Serving CPU & memory requirements above.

**Comment by mgencur** [comment]
Thanks for the fixes. They look good! And I will try to get more info about CPU/memory usage.

File: modules/serverless-troubleshooting-general.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgencur) [reply modules/serverless-troubleshooting-general.adoc:13]: Here are some details. The CPU requests are defined for each Pod in knative-serving namespace.
```
oc get pods -n knative-serving
NAME                                     READY   STATUS    RESTARTS   AGE
activator-77d49fdbc-7hgr7                2/2     Running   0          11d
activator-77d49fdbc-8whqr                2/2     Running   0          11d
autoscaler-6444dcb6c9-489jd              2/2     Running   0          11d
autoscaler-6444dcb6c9-vrldn              2/2     Running   0          11d
autoscaler-hpa-59698cf457-2npx9          2/2     Running   0          11d
autoscaler-hpa-59698cf457-ntk5p          2/2     Running   0          11d
controller-54cf5b4ccc-6bb9x              2/2     Running   0          11d
controller-54cf5b4ccc-tht92              2/2     Running   0          11d
domain-mapping-7546dfd485-4xplg          2/2     Running   0          11d
domain-mapping-7546dfd485-hkrxs          2/2     Running   0          11d
domainmapping-webhook-64784df6d7-2295s   2/2     Running   0          11d
domainmapping-webhook-64784df6d7-wjcq4   2/2     Running   0          11d
webhook-7874c59b84-56pgs                 2/2     Running   0          11d
webhook-7874c59b84-lbfns                 2/2     Running   0          11d

oc get pod -n knative-serving -ojsonpath='{.items[*].spec.containers[*].resources.requests.cpu}'
300m 10m 300m 10m 100m 10m 100m 10m 30m 10m 30m 10m 100m 10m 100m 10m 30m 10m 30m 10m 100m 10m 100m 10m 100m 10m 100m 10m

oc get pod -n knative-serving -ojsonpath='{.items[*].spec.containers[*].resources.requests.memory}'
60Mi 20Mi 60Mi 20Mi 100Mi 20Mi 100Mi 20Mi 40Mi 20Mi 40Mi 20Mi 100Mi 20Mi 100Mi 20Mi 40Mi 20Mi 40Mi 20Mi 100Mi 20Mi 100Mi 20Mi 100Mi 20Mi 100Mi 20Mi
```
Overall, this is:
CPU request: 1660m
Memory request: 1360Mi

However, we don't define CPU and memory limits for all Pods, only for some of them. So it's probably doesn't make sense to mention CPU/memory limits because they can grow beyond the number you will get.

Also, the Pods in knative-serving-ingress namespace don't define any CPU/memory requests or limits. There's a JIRA for that: https://issues.redhat.com/browse/SRVKS-532
But there are some Pods:
```
oc get pods -n knative-serving-ingress
NAME                                      READY   STATUS    RESTARTS   AGE
3scale-kourier-gateway-785c74cd65-8qb2g   1/1     Running   0          11d
3scale-kourier-gateway-785c74cd65-s7w74   1/1     Running   0          11d
net-kourier-controller-7976dbcbbb-jdpdr   1/1     Running   0          11d
net-kourier-controller-7976dbcbbb-m5qfq   1/1     Running   0          11d
```
These Pods belong to "knative-serving" component. So, when talking about CPU/memory requests we may need to say "knative-serving" component requests "at least" XYZ memory and XYZ cpu.

[the remark this answers is not in the document]
Comment (rh-max) [reply modules/serverless-troubleshooting-general.adoc]: @mgencur Perfect, I've added this information, does it look good now?
https://deploy-preview-45262--osdocs.netlify.app/openshift-enterprise/latest/serverless/reference/troubleshooting.html#serverless-troubleshooting-general_serverless-troubleshooting
Thank you.

**Review by michaelryanpeter — COMMENTED** [review]
While troubleshooting guides can diverge from the standards of concept/procedure/reference, I question the FAQ-style employed in this guide. I find it a little distracting, and difficult to apply consistently with topics you are trying to cover. I think that the Service Mesh troubleshooting guide is a great assembly to use as a model. It also adheres to a more standard implementation of the mod docs guidelines.

You have done a good job of providing troubleshooting help for requests. I wonder if the subsequent sections would be better served in an overview guide at the beginning of the book. Have you considered taking this approach?

File: serverless/reference/troubleshooting.adoc [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+This section addresses common issues with {ServerlessProductName}.
+
+// General
+include::modules/serverless-troubleshooting-general.adoc[leveloffset=+1]
+
+[role="_additional-resources"]
+[id="additional-resources_troubleshooting-general"]
+=== Additional resources
+
+* xref:../../serverless/admin_guide/serverless-configuration.html#serverless-configuration[Global configuration]
```
Comment (michaelryanpeter) [inline serverless/reference/troubleshooting.adoc:18]:
```suggestion
* xref:../../serverless/admin_guide/serverless-configuration.adoc#serverless-configuration[Global configuration]
```
Please make sure you are ending your `xrefs` with `.adoc` throughout.

**Review by michaelryanpeter — COMMENTED** [review]
Two things I forgot to mention:
1. Please remember to squash your commits before requesting a merge
2. `xrefs` need to end with `.adoc`, not `.html`. The links will work in the preview, but it is important for publishing on the portal.

File: modules/serverless-troubleshooting-general.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgencur) [reply modules/serverless-troubleshooting-general.adoc]: Originally, my intention was to only explain where the numbers are coming from and not to include them in the docs in the given format. But it doesn't look too bad. This is probably as much info as we can provide.

File: modules/serverless-troubleshooting-general.adoc [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+//
+// * serverless/reference/troubleshooting.adoc
+
+:_content-type: REFERENCE
+[id="serverless-troubleshooting-general_{context}"]
+= General troubleshooting
+
+*How many {product-title} resources does {ServerlessProductShortName} Serving consume?*
+
+Resource usage can be defined in terms of _requests_ and in terms of _limits_.
```
[the remark this answers is not in the document]
Comment (rh-max) [reply modules/serverless-troubleshooting-general.adoc:11]: It's "can be" and not "is", because there are many ways to define resource usage, and here we decided to use the specific definition of requests and limits.

File: modules/serverless-troubleshooting-general.adoc [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+----
+60Mi 20Mi 60Mi 20Mi 100Mi 20Mi 100Mi 20Mi 40Mi 20Mi 40Mi 20Mi 100Mi 20Mi 100Mi 20Mi 40Mi 20Mi 40Mi 20Mi 100Mi 20Mi 100Mi 20Mi 100Mi 20Mi 100Mi 20Mi
+----
+
+Overall, `knative-serving` amounts to these requests at the minimum:
+
+* CPU request: 1660m
+* Memory request: 1360Mi
+
+The `knative-serving-ingress` namespace, which also belongs to the Knative Serving component, does not define any CPU or memory requests. The following pods are in the `knative-serving-ingress` namespace and consume additional CPU and memory:
```
[the remark this answers is not in the document]
Comment (rh-max) [reply modules/serverless-troubleshooting-general.adoc:67]: That's correct, but even many of the native speakers of English don't know the difference between "that" and "which", so "also" is useful here.

File: modules/serverless-troubleshooting-general.adoc [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+[source,terminal]
+----
+NAME                                      READY   STATUS    RESTARTS   AGE
+3scale-kourier-gateway-785c74cd65-8qb2g   1/1     Running   0          11d
+3scale-kourier-gateway-785c74cd65-s7w74   1/1     Running   0          11d
+net-kourier-controller-7976dbcbbb-jdpdr   1/1     Running   0          11d
+net-kourier-controller-7976dbcbbb-m5qfq   1/1     Running   0          11d
+----
+
+Unlike the requests, the _limits_ are undefined for some pods in `knative-serving` and for all pods in `knative-serving-ingress`, so there is no specific limits information.
```
[the remark this answers is not in the document]
Comment (rh-max) [reply modules/serverless-troubleshooting-general.adoc:84]: I think your wording does not convey the same information as the one I've used, plus it demands a different structure of the text.

File: modules/serverless-troubleshooting-general.adoc [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+The actual CPU and memory usage of the Knative Serving component depends on the control plane and data plane load of Serving, which correlates with the amount of traffic.
+
+*How to customize {ServerlessProductName}?*
+
+Extensive configuration options are available for {ServerlessProductShortName} using config maps. For instructions, see the "Administer" section of the {ServerlessProductShortName}, beginning with "Global configuration".
```
[the remark this answers is not in the document]
Comment (rh-max) [reply modules/serverless-troubleshooting-general.adoc:90]: The optimal thing here would be to refer to the Administer section as a whole, but we can't do that. So, I'm somewhat incidentally linking to Global configuration and then explaining that the actual location of the information is from Global configuration until the end of the Administer section.

**Comment by rh-max** [comment]
Thanks @michaelryanpeter for the great review!

I implemented much of the feedback. Let me address the rest.

The goal of the guide is to address the common problems customers have. As it happens, the problems that have been reported so far (and made it into this document), are fully answered in the rest of the Serverless documentation.

Hence, while in the first iteration I did include details of how to solve the problems, I eventually removed them. That's because details in the Troubleshooting section were not as detailed as in the main docs, and they were duplicating the content anyway. That is why I explicitly made answers to the questions into bare directions to the relevant parts of the main Serverless docs.

As a result, this is different from the structure employed by the Service Mesh docs, where they provide the actual details inside the Troubleshooting section, and where the content more closely follows either the procedure or reference format.

That is why, until the set of problems addressed by Troubleshooting Serverless changes, I plan to keep it in this form.

To some more specific points:

* Some of the questions addressed in Troubleshooting Serverless are general. The reason for this is that the questions were initially very specific, but I generalized them, because the initial questions contained details that were unnecessarily specific. With questions phrased more generally, the text is cleaner and a wider range of questions is addressed now. For instance, the current "How to create a KafkaChannel or other channels" question is very general (and not even a question). It evolved from a more specific problem, the solution to which was to correctly follow the instructions on creating [Kafka]channels. Yet, if someone encounters that specific problem, the link here leads them to the solution.

* Regarding your suggestions of moving things into the Overview section. The whole troubleshooting guide could technically be boiled down to "please read the main documentation", because everything is answered there. Hence, there's no point in moving parts of Troubleshooting to the main docs (including Overview), because if the user reads the main docs thoroughly, then they don't get these problems in the first place. What we are really solving with this section is the reader not using the main docs or failing to find the information in there, and thus encountering one of the common problems.

**Comment by bergerhoffer** [comment]
The enterprise-4.12 label has been added to this PR.

This is because your PR targets the `main` branch and is labeled for enterprise-4.11. And any PR going into `main` must also target the latest version branch (enterprise-4.12).

If the update in your PR does NOT apply to version 4.12 onward, please re-target this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.

**Comment by kalexand-rh** [comment]
Only versions 4.8+ are still in maintenance. I am removing labels from earlier versions from this PR.

**Comment by bergerhoffer** [comment]
The `branch/enterprise-4.13` label has been added to this PR.

This is because your PR targets the `main` branch and is labeled for enterprise-4.12. And any PR going into `main` must also target the latest version branch (enterprise-4.13).

If the update in your PR does NOT apply to version 4.13 onward, please re-target this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.
